README.md
<div align="center">  
  <a href="https://github.com/avasuro/json-rpc-msg">  
     <p>JSON-RPC MSG</p>
  </a> 
  <p>  
    Zero-dependencies library for <a href="http://nodejs.org">Node.js</a> and <a href="https://en.wikipedia.org/wiki/JavaScript">JavaScript</a> to parse and create <a href="https://www.jsonrpc.org/specification">JSON RPC 2.0</a> messages.
  </p>
  <a href="https://img.shields.io/node/v/json-rpc-msg.svg?style=flat-square">
      <img alt="undefined" src="https://img.shields.io/node/v/json-rpc-msg.svg?style=flat-square">
  </a>
  <a href="https://img.shields.io/coveralls/github/avasuro/json-rpc-msg.svg?style=flat-square">
    <img alt="undefined" src="https://img.shields.io/coveralls/github/avasuro/json-rpc-msg.svg?style=flat-square">
  </a>
  <a href="https://img.shields.io/bundlephobia/min/json-rpc-msg.svg?style=flat-square">
      <img alt="undefined" src="https://img.shields.io/bundlephobia/min/json-rpc-msg.svg?style=flat-square">
    </a>
  <a href="https://img.shields.io/github/license/avasuro/json-rpc-msg.svg?style=flat-square">
    <img alt="undefined" src="https://img.shields.io/github/license/avasuro/json-rpc-msg.svg?style=flat-square">
  </a>
</div>

## About  
  
The **json-rpc-msg** is a library with zero dependencies that provides an API to parse and create valid JSON-RPC 2.0 messages.  
  
## Quick start  
  
Install:  
```
npm install --save json-rpc-msg 
```  

Usage:

```javascript
const JsonRPC = require('json-rpc-msg');

function sendMessageToClient(message) {
    // Some code that sends passed message to clients, e.g. over HTTP or websocket
}

/*--------------------------------------
 | Create JSON-RPC request
 |--------------------------------------
 |*/
const requestObj = JsonRPC.createRequest(123, 'updateUser', {id: 1, name: 'Alex'});
sendMessageToClient(requestObj);
// {
//   jsonrpc: '2.0',
//   id: 123,
//   method: 'updateUser',
//   params: {id: 1, name: 'Alex'}
// }

/*--------------------------------------
 | Parse JSON-RPC message 
 |--------------------------------------
 |*/
try {
    // Valid message:
    const message = JsonRPC.parseMessage('{"jsonrpc": "2.0", "id": 123, "method": "update", "params": [1,2,3]}');
    // {
    //     type: 'request',
    //     payload: {
    //         id: 123,
    //         method: 'update',
    //         params: [1,2,3]
    //     }
    // }
    
    
    // Invalid message (ParserError will be thrown):
    const failedMessage = JsonRPC.parseMessage('not valid JSON RPC message');
}
catch(e) {
    // If parser failed to parse message - send error to client:
    if (e instanceof JsonRPC.ParserError) {
        sendMessageToClient(e.rpcError);
    }
    // Otherwise throw error up:
    else {
        throw e;
    }
}
```

## API Reference

### Constants

#### ERRORS

Contains list of JSON-RPC error codes. Map with the following structure:
```javascript
JsonRPC.ERRORS = {
    SOME_ERROR: {
        code: 123,
        message: 'Some message for error'
    }
}
```

Can be used to manually create response with JSON-RPC error:
```javascript
// Create "Method not found" error for received request with ID=123
const errorObj = JsonRPC.createError(123, JsonRPC.ERRORS.METHOD_NOT_FOUND);
// {
//   jsonrpc: '2.0',
//   id: 123,
//   error: {
//     code: -32601,
//     message: 'Method not found'
//   }
// }
```

Some particular implementations of JSON-RPC protocol may want to add additional custom errors to
this object (note that errors from -32768 to 32000 are reserved for future use, for more information
see [specification to error object in JSON-RPC v2.0](https://www.jsonrpc.org/specification#error_object)):
```javascript
JsonRPC.ERRORS.NEW_ERROR_FOR_SOME_IMPLEMENTATION = {
    code: -32000,
    message: 'Message of added error'
}
const message = JsonRPC.createError(123, JsonRPC.ERRORS.NEW_ERROR_FOR_SOME_IMPLEMENTATION);
// {
//   jsonrpc: '2.0',
//   id: 123,
//   error: {
//     code: -32000,
//     message: 'Message of added error'
//   }
// }
```

#### MESSAGE_TYPES

Contains list of supported message types.

- **REQUEST** - message with "id" and "method" props (and optional "params")
- **INTERNAL_REQUEST** - the same as **REQUEST**, but only if "method" starts with "rpc.". This type of request is only for rpc-internal methods and extensions.
- **NOTIFICATION** - message without "id", and with "method" prop (and optional "params")
- **INTERNAL_NOTIFICATION** - the same as **NOTIFICATION**, but only if "method" starts with "rpc.". This type of notification is only for rpc-internal notifications and extensions.
- **RESPONSE** - message with "id" and "result" props. Represents response from server.
- **ERROR** - message with "id" and "error" props. Represents response from server with error.
- **BATCH** - message that is a group of JSON-RPC messages (array).

This constant should be mainly used to check type of parsed message:
```javascript
const receivedJSON = '{"jsonrpc": "2.0", "id": 1, "method": "log", "params": ["Some string to log"]}';
const message = JsonRPC.parseMessage(receivedJSON);
switch (message.type) {
    case JsonRPC.MESSAGE_TYPES.REQUEST:
        // Do something with request
        break;
    case JsonRPC.MESSAGE_TYPES.NOTIFICATION:
        // Do something with notification
        break;
    // ...etc
}
```

#### INTERNAL_MESSAGE_PREFIX

Contains string ("rpc."), that should be added to start of any message/notification that is intended for
internal usage of some particular JSON-RPC implementation. e.g:
```javascript
// Creates notification with name "rpc.someNotification":
const notificationJson = JsonRPC.createInternalNotification(`${JsonRPC.INTERNAL_MESSAGE_PREFIX}someNotification`);

// Because "createInternalNotification" adds this prefix automatically - the following code will
// produce the same result:
const notificationJson2 = JsonRPC.createInternalNotification('someNotification');

console.log(notificationJson === notificationJson2); // => true
```
### Methods

#### createRequest(id, method \[,params])

Creates JSON-RPC request

NOTE: If method name starts with "rpc." error will be thrown. To create requests for "rpc." methods use [createInternalRequest](#createInternalRequest) function instead.

- ``id`` {string|number}: request ID
- ``method`` {string}: method name to call
- ``params`` {Object|Array} _(optional)_: parameters to pass in method

```javascript
const requestObj = JsonRPC.createRequest(123, 'refreshAllUsers')
// {
//   jsonrpc: '2.0',
//   id: 123,
//   method: 'refreshAllUsers',
// }
```

```javascript
const requestObj = JsonRPC.createRequest(123, 'updateUser', {id: 1, name: 'Alex'})
// {
//   jsonrpc: '2.0',
//   id: 123,
//   method: 'updateUser',
//   params: {id: 1, name: 'Alex'}
// }
```

#### createNotification(method, \[,params])

Creates JSON-RPC notification

NOTE: If notification name starts with "rpc." error will be thrown. To create requests for "rpc." notifications use [createInternalNotification](#createInternalNotification) function instead.

- ``method`` {string}: method name to call
- ``params`` {Object|Array} _(optional)_: parameters to pass in method

```javascript
const notificationObj = JsonRPC.createNotification('allUsersWereRefreshed')
// {
//   jsonrpc: '2.0',
//   method: 'allUsersWereRefreshed'
// }
```

```javascript
const notificationObj = JsonRPC.createNotification('userUpdated', {id: 1, name: 'Alex'})
// {
//   jsonrpc: '2.0',
//   method: 'userUpdated',
//   params: {id: 1, name: 'Alex'}
// }
```

#### createResponse(id, result)

Creates successfull JSON-RPC response to some request

- ``id`` {string|number}: request ID
- ``result`` {Mixed}: response data

```javascript
const responseObj = JsonRPC.createResponse(123, {id: 1, name: 'Alex', age: 23})
// {
//   jsonrpc: '2.0',
//   id: 123,
//   response: {id: 1, name: 'Alex', age: 23}
// }
```

```javascript
const responseObj = JsonRPC.createResponse(123, null)
// {
//   jsonrpc: '2.0',
//   id: 123,
//   response: null 
// }
```

#### createError(id, error, \[, additionalErrorData])

Creates JSON-RPC response with error 

- ``id`` {string|number}: request ID
- ``error`` {number|{code: number, message: string?}}: error code (number) or object with "code" and optional "message" property
- ``additionalErrorData`` {Object|Array}: additional data for error

```javascript
// All of the following statements will produce same result:
const errorResponseObj = JsonRPC.createError(123, -32602);
const errorResponseObj = JsonRPC.createError(123, JsonRPC.ERRORS.INVALID_PARAMS);
const errorResponseObj = JsonRPC.createError(123, {code: -32602});
const errorResponseObj = JsonRPC.createError(123, {code: -32602, message: 'Invalid params'});
// {
//   jsonrpc: '2.0',
//   id: 123,
//   error: {
//     code: -32602,
//     message: 'Invalid params'
//   }
// }
```

There is ability to provide additional information for error, passing _additionalParams_ argument,
e.g. if invalid user name and age where passed in request:
```javascript
const errorResponseObj = JsonRPC.createError(123, JsonRPC.ERRORS.INVALID_PARAMS, {
    name: "User name can't be less then 6 characters",
    age: "User age is not defined"
});
// {
//   jsonrpc: '2.0',
//   id: 123,
//   error: {
//     code: -32602,
//     message: 'Invalid params',
//     data: {
//         name: 'User name can't be less then 6 characters',
//         age: 'User age is not defined'
//     }
//   }
// }
```

If ``error`` passed as object with property "code" but without property "message" this method will
look at JsonRPC.ERRORS to find error with given code, and find appropriate message. If there is no
error with given code in ERRORS - message will be defaulted to "Internal Server Error":
```javascript
JsonRPC.ERRORS.NEW_ERROR = {
    code: 666,
    message: 'Number of the Beast'
}

// Message will be taken from ERRORS object:
const errorResponseObj = JsonRPC.createError(123, {code: 666});
// {
//   jsonrpc: '2.0',
//   id: 123,
//   error: {
//     code: 666,
//     message: 'Number of the Beast'
// }

// Message in ERRORS object will not be used because "message" is provided:
const errorResponseObj = JsonRPC.createError(123, {code: 666, message: 'replaced message'});
// {
//   jsonrpc: '2.0',
//   id: 123,
//   error: {
//     code: 666,
//     message: 'replaced message'
// }

// Error that is not in ERRORS object will have default "Internal Serve Error" message:
const errorResponseObj = JsonRPC.createError(123, {code: 999});
// {
//   jsonrpc: '2.0',
//   id: 123,
//   error: {
//     code: 999,
//     message: 'Internal Server Error'
// }
```

#### parseMessage(message)

Validates and parses given JSON-RPC message

- ``message`` {string|Object|array} - JSON string or message object or batch of messages (array)

Returns object with properties "type" and "payload". Type can be one of
the types listed in JsonRPC.MESSAGE_TYPES constant. Payload will contain message body, which depends
on message type:
- **REQUEST** and **INTERNAL_REQUEST** will contain "id", "method" and optional "params" properties:
    ```javascript
    const message = JsonRPC.parseMessage('{"jsonrpc": "2.0", "id": 123, "method": "update", "params": [1,2,3]}');
    // {
    //     type: 'request',
    //     payload: {
    //         id: 123,
    //         method: 'update',
    //         params: [1,2,3]
    //     }
    // }
    ```
- **NOTIFICATION** and **INTERNAL_NOTIFICATION** will contain "method" and optional "params" properties:
    ```javascript
    const message = JsonRPC.parseMessage('{"jsonrpc": "2.0", "method": "updated", "params": [1,2,3]}');
    // {
    //     type: 'notification',
    //     payload: {
    //         method: 'updated',
    //         params: [1,2,3]
    //     }
    // }
    ```
- **RESPONSE** will contain "id" and "result" properties:
    ```javascript
    const message = JsonRPC.parseMessage('{"jsonrpc": "2.0", "id": 1, "result": [1,2,3]}');
    // {
    //     type: 'response',
    //     payload: {
    //         id: 1,
    //         result: [1,2,3]
    //     }
    // }
    ```
- **ERROR** will contain "id" and "error" properties:
    ```javascript
    const message = JsonRPC.parseMessage('{"jsonrpc": "2.0", "id": 1, "error": {"code": -32602, "message": "Invalid params", "data": [1,2,3]}}');
    // {
    //     type: 'error',
    //     payload: {
    //         id: 1,
    //         error: {
    //             code: -32602,
    //             message: 'Invalid params',
    //             data: [1,2,3] 
    //         }
    //     }
    // }
    ```
- **BATCH** will contain array of parsed messages in "payload" property. If some of this messages failed to be parsed - there will be an error in this array instead of parsed message:
    ```javascript
    const message = JsonRPC.parseMessage('['+
        '{"jsonrpc": "2.0", "id": 1, "result": "OK"}' +
        '{"jsonrpc": "2.0", "id": 1, "error": {"code": -32602, "message": "Invalid params", "data": [1,2,3]}}]' +
        'invalid JSON-RPC message' +
    ']');
    // {
    //   type: 'batch',
    //   payload: [
    //     {
    //       type: 'response',
    //       payload: {
    //         id: 1,
    //           result: [1,2,3]
    //       }
    //     },
    //     {
    //       type: 'error',
    //       payload: {
    //         id: 1,
    //         error: {
    //           code: -32602,
    //           message: 'Invalid params',
    //           data: [1,2,3] 
    //         }
    //       },
    //     },
    //     [ParserError Object]
    //   ] 
    // }
    ```

#### createInternalRequest

The same as [createRequest](#createRequest), but allows to create requests with method name
that starts with "rpc." prefix. If this prefix is not presented in method name it will be
added automatically.

#### createInternalNotification

The same as [createNotification](#createNotification), but allows to create notification with name
that starts with "rpc." prefix. If this prefix is not presented in name it will be
added automatically.

## License  
  
  [MIT](LICENSE)

CHANGELOG.md
1.1.1 / 2019-07-16
==================
 - [dev] removed: package-lock.json and prevented from creating
 - [dev] updated: eslint config + fixed eslint errors
 - [dev] added: run code lint before commit
 - [added] changelog file
 
1.1.0 / 2019-01-16
==================
 - [fix] [#parseMessage] throws an error if empty batch passed (e.g. jsonrpc.parseMessage([]))
 - [fix] [#parseMessage] throws an error if request parameters have invalid type (any type, except "object")
 - [fix] [#parseMessage] throws an error if notification/request method is not defined of has invalid type (any type, except "string")
 - [added] [#parseMessage] validation of error object. Now if error have no code or message new ParserError will be thrown. Also this error will be thrown if code, message or error object itself has invalid type.
 - [fix] [#createRequest] Treat request parameters that equals to "null" as not defined
 - [fix] [#createInternalRequest] Treat request parameters that equals to "null" as not defined
 - [fix] [#createNotification] Treat request parameters that equals to "null" as not defined
 - [fix] [#createInternalNotification] Treat request parameters that equals to "null" as not defined
 - [improvement] [#createError] Additional error data now can be of any type
 - [fix] By a mistake ParserError was stored JSON-RPC encoded error inside errorData property. Now it's moved directly into error object.
 
 1.0.0 / 2019-01-15
 ==================
  - initial implementation

test/createError.spec.js
const JsonRPC = require('../src/json-rpc-parse');
const expect = require('chai').expect;

const RPC_VERSION = process.env.RPC_VERSION;

describe('#createError', () => {
    function createErrorWithCode(error) {
        return () => JsonRPC.createError(1, error);
    }

    function createErrorWithId(id) {
        return () => JsonRPC.createError(id, 123);
    }

    function createErrorWithAdditionalData(data) {
        return () => JsonRPC.createError(1, 123, data);
    }

    it('Returns valid error object', () => {
        function testError(code, additionalData) {
            const requestId = 1;
            const result = JsonRPC.createError(requestId, code, additionalData);
            expect(result).to.be.an('object');
            expect(result).to.have.property('jsonrpc', RPC_VERSION);
            expect(result).to.have.property('id', requestId);
            expect(result).to.have.deep.property('error', {
                code: 123,
                message: 'Internal Server Error',
                data: additionalData
            });
            expect(result).to.not.have.property('result');
        }
        testError(123, {someData: 'test'});
    });

    it('Allows to set request ID as null, string or number', () => {
        expect(createErrorWithId(null)).to.not.throw(Error);
        expect(createErrorWithId('123')).to.not.throw(Error);
        expect(createErrorWithId(123)).to.not.throw(Error);
    });

    it('Throws an error if request ID is not defined', () => {
        expect(createErrorWithId(undefined)).to.throw(Error);
        expect(createErrorWithId('')).to.throw(Error);
    });

    it('Throws an error if request ID have incorrect type', () => {
        expect(createErrorWithId(1.2)).to.throw(Error);
        expect(createErrorWithId(true)).to.throw(Error);
        expect(createErrorWithId(false)).to.throw(Error);
        expect(createErrorWithId({})).to.throw(Error);
        expect(createErrorWithId([])).to.throw(Error);
    });

    it('Allows to set error code as integer number', () => {
        expect(createErrorWithCode(123)).to.not.throw(Error);
        expect(createErrorWithCode(0)).to.not.throw(Error);
    });

    it('Allows to pass error code as object, with "code" and optional "message" property', () => {
        expect(createErrorWithCode({code: 123, message: 'Some custom error'})).to.not.throw(Error);
        expect(createErrorWithCode({code: JsonRPC.ERRORS.INTERNAL_ERROR.code})).to.not.throw(Error);
        expect(createErrorWithCode(JsonRPC.ERRORS.INTERNAL_ERROR)).to.not.throw(Error);
    });

    it('Throws an error if error code is not passed', () => {
        expect(createErrorWithCode(undefined)).to.throw(Error);
        expect(createErrorWithCode(null)).to.throw(Error);
        expect(createErrorWithCode('')).to.throw(Error);
        expect(createErrorWithCode({message: 'Some custom error'})).to.throw(Error);
        expect(createErrorWithCode({})).to.throw(Error);
    });

    it('Throws an error if error code has invalid type', () => {
        expect(createErrorWithCode('123')).to.throw(Error);
        expect(createErrorWithCode(true)).to.throw(Error);
        expect(createErrorWithCode(false)).to.throw(Error);
        expect(createErrorWithCode([])).to.throw(Error);
        expect(createErrorWithCode(123.456)).to.throw(Error);
        expect(createErrorWithCode({code: '123'})).to.throw(Error);
        expect(createErrorWithCode({code: '123', message: 'test'})).to.throw(Error);
    });

    it('Allows to set additional error data as Structured Value (array or object) or as Primitive', () => {
        expect(createErrorWithAdditionalData([])).to.not.throw(Error);
        expect(createErrorWithAdditionalData({})).to.not.throw(Error);
        expect(createErrorWithAdditionalData(0)).to.not.throw(Error);
        expect(createErrorWithAdditionalData(1)).to.not.throw(Error);
        expect(createErrorWithAdditionalData(1.25)).to.not.throw(Error);
        expect(createErrorWithAdditionalData(false)).to.not.throw(Error);
        expect(createErrorWithAdditionalData(true)).to.not.throw(Error);
        expect(createErrorWithAdditionalData('123')).to.not.throw(Error);
        expect(createErrorWithAdditionalData('')).to.not.throw(Error);
    });

    it('Allows to skip additional error data', () => {
        expect(createErrorWithAdditionalData(undefined)).to.not.throw(Error);
        expect(createErrorWithAdditionalData(null)).to.not.throw(Error);
    });
});


src/json-rpc-parse.js
((factory) => {
    // Define as CommonJS export:
    if (typeof require === 'function' && typeof exports === 'object') {
        module.exports = factory(); // eslint-disable-line
    }
    // Define as AMD:
    else if (typeof define === 'function' && define.amd) { // eslint-disable-line no-undef
        define([], factory); // eslint-disable-line no-undef
    }
    // Browser:
    else {
        window.JsonRPC = factory(); // eslint-disable-line no-undef
    }
})(() => {
    /**
     * List of errors
     *
     * @type {object}
     */
    const ERRORS = {
        // codes from -32000 to -32099 reserved for implementation-defined server-errors.
        INVALID_REQUEST: {
            code: -32600,
            message: 'Invalid Request'
        },
        METHOD_NOT_FOUND: {
            code: -32601,
            message: 'Method not found'
        },
        INVALID_PARAMS: {
            code: -32602,
            message: 'Invalid params'
        },
        INTERNAL_ERROR: {
            code: -32603,
            message: 'Internal error'
        },
        PARSE_ERROR: {
            code: -32700,
            message: 'Parse error'
        }
    };

    /**
     * List of supported message types
     *
     * @type {object}
     */
    const MESSAGE_TYPES = {
        REQUEST: 'request',
        INTERNAL_REQUEST: 'internal_request',
        NOTIFICATION: 'notification',
        INTERNAL_NOTIFICATION: 'internal_notification',
        RESPONSE: 'response',
        ERROR: 'error',
        BATCH: 'batch'
    };

    /**
     * Contains prefix of name of all messages for internal usage
     *
     * @type {string}
     */
    const INTERNAL_MESSAGE_PREFIX = 'rpc.';

    /**
     * Constructor of error object, that should be thrown from "parse" method
     *
     * @param {object} error
     *
     * @constructor
     */
    function ParserError(error) {
        this.name = 'ParserError';
        this.message = 'Failed to parse JSON-RPC message';
        this.rpcError = error;
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, ParserError);
        }
        else {
            this.stack = (new Error()).stack;
        }
    }

    ParserError.prototype = Object.create(Error.prototype);
    ParserError.prototype.constructor = ParserError;

    /**
     * Utility function for assertion
     *
     * @param {boolean} condition
     * @param {string} errorText
     *
     * @returns {void}
     *
     * @throws Error if condition is falsy
     */
    function assert(condition, errorText) {
        if (!condition) {
            throw new Error(errorText);
        }
    }

    /**
     * Utility function to create basic JSON-RPC message
     *
     * @param {number|string|undefined} id?
     * @param {object} messageData?
     *
     * @returns {object}
     *
     * @private
     */
    function _createMessage(id, messageData = null) {
        if (id !== undefined && id !== null) {
            if (typeof id === 'string') {
                assert(id.length > 0, 'ID of message is not defined');
            }
            else {
                assert(Number.isInteger(id), `ID should be a string or an integer number, "${typeof id}" given`);
            }
        }

        const message = {
            jsonrpc: '2.0',
            ...messageData
        };

        if (id !== undefined) {
            message.id = id;
        }

        return message;
    }

    /**
     * Create a JSON-RPC request object ( https://www.jsonrpc.org/specification#request_object )
     *
     * @param {number|string|undefined} id
     * @param {string} method
     * @param {object|array|undefined} params
     * @param {boolean} isInternal
     *
     * @returns {object}
     *
     * @private
     */
    function _createRequestObject(id, method, params, isInternal) {
        assert(typeof method === 'string', 'Method should be a string');
        assert(method.trim().length > 0, 'Method name should not be empty');
        if (!isInternal) {
            assert(method.indexOf(INTERNAL_MESSAGE_PREFIX) !== 0, `Invalid method name "${method}": only internal methods can be prefixed with "${INTERNAL_MESSAGE_PREFIX}"`);
        }
        if (params !== undefined && params !== null) {
            assert(
                Array.isArray(params) || typeof params === 'object',
                `Request parameters must be a structured value (an array or an object), "${typeof params}" given`
            );
        }

        if (isInternal) {
            if (method.indexOf(INTERNAL_MESSAGE_PREFIX) !== 0) {
                method = `${INTERNAL_MESSAGE_PREFIX}${method}`;
            }
        }

        let messageData = {method};
        if (params) {
            messageData.params = params;
        }

        return _createMessage(id, messageData);
    }

    /**
     * Utility function to create JSON-RPC request
     *
     * @param {number|string} id
     * @param {string} method
     * @param {object|array|undefined} params
     * @param {boolean} isInternal
     *
     * @returns {object}
     *
     * @private
     */
    function _createRequest(id, method, params, isInternal) {
        assert(id !== undefined && id !== null, 'ID of request is not defined');
        return _createRequestObject(id, method, params, isInternal);
    }

    /**
     * Utility function to create JSON-RPC notification
     *
     * @param {string} method
     * @param {object|array|undefined} params
     * @param {boolean} isInternal
     *
     * @returns {object}
     *
     * @private
     */
    function _createNotification(method, params, isInternal) {
        return _createRequestObject(undefined, method, params, isInternal);
    }

    /**
     * Utility function to create JSON-RPC response
     *
     * @param {number|string} id
     * @param {object} responseData
     *
     * @returns {object}
     *
     * @private
     */
    function _createResponse(id, responseData) {
        assert(id !== undefined, 'ID of response is not defined');
        return _createMessage(id, responseData);
    }

    /**
     * Creates a response error object ( https://www.jsonrpc.org/specification#error_object )
     *
     * @param {Number|{code: int, message: string?}} errorData - error code
     * @param {*} additionalData? - error details
     *
     * @return {Object}
     */
    function _createErrorObject(errorData, additionalData) {
        if (Number.isInteger(errorData)) {
            errorData = {code: errorData};
        }
        assert(
            typeof errorData === 'object' && errorData !== null && !Array.isArray(errorData),
            'Error code should be an integer number or object with "code" and "message" property'
        );
        assert(errorData.message === undefined || typeof errorData.message === 'string', 'Error message should be a string');
        assert(Number.isInteger(errorData.code), 'Error code should be an integer value');

        if (!errorData.message) {
            let errorInRegistry = Object.values(ERRORS).find(
                error => error.code === errorData.code
            );
            errorData.message = errorInRegistry ? errorInRegistry.message : 'Internal Server Error';
        }

        const errorObj = {...errorData};
        if (additionalData !== null && additionalData !== undefined) {
            errorObj.data = additionalData;
        }

        return errorObj;
    }

    /**
     * Creates new JSON-RPC request
     *
     * @param {number|string} id
     * @param {string} method
     * @param {object|array|undefined} params
     *
     * @returns {object}
     */
    function createRequest(id, method, params) {
        return _createRequest(id, method, params, false);
    }

    /**
     * Creates new internal JSON-RPC request
     *
     * @param {number|string} id
     * @param {string} method
     * @param {object|array|undefined} params
     *
     * @returns {object}
     */
    function createInternalRequest(id, method, params) {
        return _createRequest(id, method, params, true);
    }

    /**
     * Creates new JSON-RPC notification
     *
     * @param {string} method
     * @param {object|array|undefined} params
     *
     * @returns {object}
     */
    function createNotification(method, params) {
        return _createNotification(method, params, false);
    }

    /**
     * Creates new internal JSON-RPC notification
     *
     * @param {string} method
     * @param {object|array|undefined} params
     *
     * @returns {object}
     */
    function createInternalNotification(method, params) {
        return _createNotification(method, params, true);
    }

    /**
     * Creates JSON-RPC response
     *
     * @param {number|string} id
     * @param {*} result
     *
     * @returns {object}
     *
     * @private
     */
    function createResponse(id, result) {
        assert(id !== undefined && id !== null, 'Response ID is not defined');
        assert(result !== undefined, 'Response result is not defined');
        return _createResponse(id, {result});
    }

    /**
     * Creates JSON-RPC response with error
     *
     * @param {number|string} id
     * @param {number|{code: number, message: string?}} errorCode
     * @param {object} details?
     *
     * @returns {object}
     */
    function createError(id, errorCode, details) {
        return _createResponse(id, {error: _createErrorObject(errorCode, details)});
    }

    /**
     * Parses passed message
     *
     * @param {string|object|array} message - JSON string or object or array or requests (batch)
     *
     * @returns {Array|object}
     */
    function parseMessage(message) {
        let result;
        // Prase JSON string into object:
        if (typeof message === 'string') {
            try {
                message = JSON.parse(message);
            }
            catch (e) {
                throw new ParserError(createError(null, ERRORS.PARSE_ERROR));
            }
        }


        // If message is an array then it's a batch, so we need to parse every request in this
        // batch separately
        // (unless we already parsing a batch - in that case parse error should be returned):
        if (Array.isArray(message)) {
            if (message.length === 0) {
                throw new ParserError(createError(null, ERRORS.INVALID_REQUEST));
            }
            result = {
                type: MESSAGE_TYPES.BATCH,
                payload: message.map((request) => {
                    let msg;
                    try {
                        msg = parseMessage(request, false);
                    }
                    catch (e) {
                        if (e instanceof ParserError) {
                            msg = e;
                        }
                    }
                    return msg;
                })
            };
        }
        // If message is normal js object, so we need to validate it and determine it's type:
        else if (typeof message === 'object') {
            // The only supported version is JSON-RPC 2.0:
            if (message.jsonrpc !== '2.0') {
                let id = message.id || null;
                throw new ParserError(createError(id, ERRORS.INTERNAL_ERROR));
            }

            if (message.hasOwnProperty('method')) {
                if (message.hasOwnProperty('params')) {
                    if (typeof message.params !== 'object' || message.params === null) {
                        throw new ParserError(
                            createError(message.id || null, ERRORS.INVALID_REQUEST)
                        );
                    }
                }

                if (typeof message.method !== 'string' || message.method.trim().length === 0) {
                    throw new ParserError(
                        createError(message.id || null, ERRORS.INVALID_REQUEST)
                    );
                }

                if (message.id) {
                    result = {
                        type: message.method.indexOf(INTERNAL_MESSAGE_PREFIX) === 0
                            ? MESSAGE_TYPES.INTERNAL_REQUEST
                            : MESSAGE_TYPES.REQUEST,
                        payload: {
                            id: message.id,
                            method: message.method,
                            params: message.params
                        }
                    };
                }
                else {
                    result = {
                        type: message.method.indexOf(INTERNAL_MESSAGE_PREFIX) === 0
                            ? MESSAGE_TYPES.INTERNAL_NOTIFICATION
                            : MESSAGE_TYPES.NOTIFICATION,
                        payload: {
                            method: message.method,
                            params: message.params
                        }
                    };
                }
            }
            else {
                if (!message.id) {
                    throw new ParserError(createError(null, ERRORS.INVALID_REQUEST));
                }

                if (
                    (message.result && message.error) ||
                    (!message.result && !message.error)
                ) {
                    throw new ParserError(createError(message.id, ERRORS.INVALID_REQUEST));
                }
                else if (message.result) {
                    result = {
                        type: MESSAGE_TYPES.RESPONSE,
                        payload: {
                            id: message.id,
                            result: message.result
                        }
                    };
                }
                else {
                    if (typeof message.error !== 'object' || message.error === null || Array.isArray(message.error)) {
                        throw new ParserError(createError(message.id, ERRORS.INVALID_REQUEST));
                    }
                    if (!message.error.hasOwnProperty('code') || !Number.isInteger(message.error.code)) {
                        throw new ParserError(createError(message.id, ERRORS.INVALID_REQUEST));
                    }
                    if (!message.error.hasOwnProperty('message') || typeof message.error.message !== 'string') {
                        throw new ParserError(createError(message.id, ERRORS.INVALID_REQUEST));
                    }
                    result = {
                        type: MESSAGE_TYPES.ERROR,
                        payload: {
                            id: message.id,
                            error: message.error
                        }
                    };
                }
            }
        }
        // If message is not an array and is not an object - it is invalid, so return an error:
        else {
            throw new ParserError(createError(null, ERRORS.INVALID_REQUEST));
        }

        return result;
    }

    return {
        ERRORS,
        MESSAGE_TYPES,
        INTERNAL_MESSAGE_PREFIX,
        ParserError,
        createRequest,
        createInternalRequest,
        createNotification,
        createInternalNotification,
        createResponse,
        createError,
        parseMessage
    };
});
